cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -149,11 +149,16 @@ private newtype TDefOrUseImpl =
 private predicate isGlobalUse(
   GlobalLikeVariable v, IRFunction f, int indirection, int indirectionIndex
 ) {
-  exists(VariableAddressInstruction vai |
-    vai.getEnclosingIRFunction() = f and
-    vai.getAstVariable() = v and
-    isDef(_, _, _, vai, indirection, indirectionIndex)
-  )
+  // Generate a "global use" at the end of the function body if there's a
+  // direct definition somewhere in the body of the function
+  indirection =
+    min(int cand, VariableAddressInstruction vai |
+      vai.getEnclosingIRFunction() = f and
+      vai.getAstVariable() = v and
+      isDef(_, _, _, vai, cand, indirectionIndex)
+    |
+      cand
+    )
```

Comment:
I'm curious why only definitions are counted as a "global use" here and not, well, uses as well.  Are they handled somewhere else?

Reply:
"global use" doesn't actually mean that there's a use of a global variable in the source code. Rather, the "global use" is a synthetic use we generate to model the value of the global variable as we're about to leave a function. Consider, for example, this snippet:
```cpp
int global;
int source();
void sink(int);

void set_global() {
  global = source();
}

void read_global() {
  sink(global);
}
```
We insert a "global use" (i.e., an SSA read of `global`) at the end of each function that writes to the global variable. We then insert a [jumpStep](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll#L699) from that global use to the dataflow node representing the global variable.

Completely dual to the notion of "global uses" is the notion of a "global def". A global def is a synthetic SSA definition of a global variable that is inserted at the _entry_ of every function that reads a global variable. So in the above example, `read_global` has a "global def" at the entry of the function that models the value of the `global` variable as we're entering the function.

So from the point-of-view of dataflow, the above program really looks like:
```cpp
int global; // (1)
int source();
void sink(int);

void set_global() {
  global = source();
  __synthetic_use(global);
}

void read_global() {
  global = __synthetic_def;
  sink(global); // (2)
}
```

and the way we get dataflow from `source` to `sink` is:
1. Flow from `source()` to `__synthetic_use` (via SSA)
2. Flow from `__synthetic_use` to `// (1)` (via a jumpStep)
3. Flow from `// (1)` the  to `__synthetic_def` (via a jumpStep)
4. Flow from `__synthetic_def` to `// (2)` (via SSA)

This roundabout way of doing flow from the `__synthetic_use` to `// (1)` and then from `// (1)` to `__synthetic_def` is a way to avoid a combinatorial explosion if there are many definitions _and_ many uses of a global variable. If we jumped directly from `__synthetic_use` to `__synthetic_def` we'd get a cartesian product between the two sets (i.e., if there were `N` defs and `M` uses we'd get a relation of size `N * M`), but by going to this intermediary we're only getting `N + M` entries in the table (i.e., `N` entries representing the flow from all `__synthetic_use`s to `// (1)`, and `M` entries from `// (1)` to all `__synthetic_def`s).

Finally, to answer this part of your question:
> Are they handled somewhere else?

All uses of variables that appear in the source code (whether they're global or not) is handled in [isUse](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll#L832) 🙂

Reply:
Thanks, this is a great explanation.  I had forgotten some of the details, and been confused because the "use" in "global use" refers to the synthetic *being* a use, not being *for* a use.

Is the architecture you described above documented anywhere?

Reply:
The trick of reducing the size of from `N * M` to `N + M` is document here: https://github.com/github/codeql/blob/main/docs/ql-libraries/dataflow/dataflow.md#flow-through-global-variables.

Reply:
I think something in the code explaining the use of synthetics for globals would be good.  It may already exist, I only briefly looked.

Reply:
Sure, I can add that 👍

Reply:
Your example code above was really good, consider using that!

Reply:
I've added a version of that example in 34980bfe2066fea0f4b7c275c23d090d9cae312e

Reply:
The new doc comment is great, thanks for doing this!